scenario/state.py

```diff
@@ -978,6 +974,35 @@ def autoload(charm_type: Type["CharmType"]):
         if actions_path.exists():
             actions = yaml.safe_load(actions_path.open())
 
+        return meta, config, actions
+
+    @staticmethod
+    def _load_metadata(charm_root: Path):
+        metadata_path = charm_root / "charmcraft.yaml"
+        meta = yaml.safe_load(metadata_path.open()) if metadata_path.exists() else {}
+        config = meta.get("config", None)
+        actions = meta.get("actions", None)
+        return meta, config, actions
+
+    @staticmethod
+    def autoload(charm_type: Type["CharmType"]):
+        charm_source_path = Path(inspect.getfile(charm_type))
+        charm_root = charm_source_path.parent.parent
+
+        metadata_path = charm_root / "metadata.yaml"
+
+        if metadata_path.exists():
```

Comment:
It feels a bit odd that the check for existence is inside `_load_metadata` but outside `_load_metadata_legacy`. Maybe this is unimportant because the legacy one might go away eventually? Maybe it doesn't matter because they're internal methods that probably wouldn't be used elsewhere?

Reply:
mmm no, that's a bug because if neither charmcraft nor metadata yamls are found, we want to raise MetadataNotFoundError and not FileNotFoundError

Reply:
added some more tests for various combinations